src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -906,10 +906,28 @@ private bool IsPossibleAttributeDeclaration()
 
         private SyntaxList<AttributeListSyntax> ParseAttributeDeclarations(bool inExpressionContext)
         {
-            var attributes = _pool.Allocate<AttributeListSyntax>();
             var saveTerm = _termState;
             _termState |= TerminatorState.IsAttributeDeclarationTerminator;
 
+            // An attribute can never appear *inside* an attribute declaration in the language.  However, during parsing
+            // we can end up in a state where we're trying to exactly that, through a path of
+            // Attribute->Argument->Expression->Attribute (since attributes can not be on lambda expressions).
+            //
+            // Worse, when we are in a deeply ambiguous (or gibberish) scenario, where we see lots of code with
+            // `... [ ... [ ... ] ... ] ...` in it, we can get into exponential speculative parsing blowup.  Where we
+            // try `[ ... ]` both as an attribute *and* a collection expression (including needing to speculate on the
+            // nested `[ ... ]`s we run into.
+            //
+            // This speculative blowup can be very bad, effectively live-locking a host while each successive, nested,
+            // speculative parse gets slower and slower.  Since we cannot ever legally have an attribute within an
+            // attribute, we bail out here immediately syntactically.  This does mean we won't parse something like:
+            // `[X([Y]() => {})]` without errors.  But as that's not semantically legal anyways, it's better to not
+            // parse that, if it means we don't blowup on code that is can't have such a detriment on the compiler
+            // experience.
```

Comment:
Consider simplifying:
```suggestion
            // Since we cannot ever legally have an attribute within an
            // attribute, we bail out here immediately syntactically.  This does mean we won't parse something like:
            // `[X([Y]() => {})]` without errors, but that is not semantically legal anyway.
```